llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp

```diff
@@ -230,6 +236,44 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
     return false;
   }
 
+  bool evaluateInstruction(const MCInst &Inst, uint64_t Addr, uint64_t Size,
+                      uint64_t &Target) const override {
+    switch(Inst.getOpcode()) {
+      default:
+        return false;
+      case RISCV::ADDI:
+      case RISCV::ADDIW: {
+        if (auto TargetRegState = getGPRState(Inst.getOperand(1).getReg())) {
+          Target  = *TargetRegState + Inst.getOperand(2).getImm();
```

Comment:
ADDIW sign extends the result.

Reply:
The sign extension is handled internally since the the immediate value has type `int64_t` which according to https://en.cppreference.com/w/cpp/types/integer is guaranteed to use 2s complement with no padding, which is essentially what sign extension does. I went ahead and printed the bit representation of the immediate values passed to the instructions and they were sign extended as per specification.

Reply:
ADDIW needs to sign extend the addition result.

Reply:
I am still a bit confused by this feedback, mainly because to my understanding, the sign extension of the result is being handled by the fact that the values are loaded in memory already sign extended to 64 bits along with 2s complement addition preserving the sign extension over arithmetic manipulation.

Here is my understanding so far:
1. A signed value, extended to 64 bits, is loaded into `*TargetRegState` by either AUIPC or LUI
2. The immediate value, also already sign extended to 64 bits, is added to `*TargetRegState`
3. 2s complement addition is carried out. The sum will stay sign extended to 64 bits as per 2s complement

Is this correct or am I missing something?

Reply:
ADDIW adds bit 31:0 of both inputs, calculates a 32 bit result then sign extends that to 64 bits. Bits 63:32 of the inputs are discarded and don't affect the result. This is different than adding 2 int64_t values together.